pubsub/event.go

```diff
@@ -0,0 +1,307 @@
+func (f Field) IsZero() bool { return !f.valid }
+
+// String returns the field value as a string.
+func (f Field) String() string { return f.value }
+
+// Bytes returns the field value as a byte slice.
+func (f Field) Bytes() []byte { return []byte(f.value) }
+
+// Int attempts to parse and return the field value as an integer.
+func (f Field) Int() (int, error) {
+	x, err := strconv.ParseInt(f.value, 10, 32)
+	return int(x), err
+}
+
+// Uint64 attempts to parse and return the field value as a uint64.
+func (f Field) Uint64() (uint64, error) { return strconv.ParseUint(f.value, 10, 64) }
+
+// Int64 attempts to parse and return the field value as a int64.
+func (f Field) Int64() (int64, error) { return strconv.ParseInt(f.value, 10, 64) }
+
+// An Event is passed to an Emitter to manage which
+// events a Listener is subscribed to.
+type Event struct {
+	fields []Field
+	kind   EventType
+}
+
+// As sets the alias of the last field in the event list. You may then call
+// Event.Find(alias) to look up the value of the field in the event.
+func (e Event) As(alias string) Event {
+	e.fields[len(e.fields)-1].alias = alias
+	return e
+}
+
+// String creates a Field containing a string.
+func (e Event) String(str string) Event {
+	e.fields = append(e.fields, Field{valid: true, value: str})
+	return e
+}
+
+// Int creates a Field containing an integer.
+func (e Event) Int(x int) Event {
+	e.fields = append(e.fields, Field{valid: true, value: strconv.Itoa(x)})
+	return e
+}
+
+// Star creates a Field containing the Kleene star `*` for pattern subscription,
+// and chains it on to the Event.
+func (e Event) Star() Event {
+	e.assertPattern()
+	e.fields = append(e.fields, Field{valid: true, value: "*", pattern: patternStar})
+	return e
+}
+
+// Placeholder creates a field containing a `?` for a placeholder in Redis patterns,
+// and chains it on to the event.
+func (e Event) Placeholder() Event {
+	e.assertPattern()
+	e.fields = append(e.fields, Field{valid: true, value: "?", pattern: patternPlaceholder})
+	return e
+}
+
+// Alternatives creates a field with the alts wrapped in brackets, to match
+// one of them in a Redis pattern, and chains it on to the event.
+func (e Event) Alternatives(alts string) Event {
+	e.assertPattern()
+	e.fields = append(e.fields, Field{
+		valid:   true,
+		value:   "[" + alts + "]",
+		pattern: patternAlts,
+	})
+	return e
+}
+
+// assertPattern panics if the event is not a Pattern type.
+func (e Event) assertPattern() {
+	if e.kind != PatternEvent {
+		panic("That operation is only valid on pattern events created with NewPattern()")
+	}
+}
+
+// Len returns the number of fields contained in the event.
+func (e Event) Len() int { return len(e.fields) }
+
+// Get returns the value of a field at index `i` within the event. If the
+// field does not exist, an empty struct will be returned.
+func (e Event) Get(i int) Field {
+	if len(e.fields) <= i {
+		return Field{valid: false}
```

Comment:
This is not entirely an accurate representation of what's happening with `Get(out_of_bounds)`. I would change the signature to return a `*Field` and pass `nil` here instead.